alf/utils/conditional_ops.py

```diff
@@ -62,58 +61,29 @@ def conditional_update(target, cond, func, *args, **kwargs):
     Returns:
         nest with the same structure and shape as target.
     """
-    # shape of indices from where() is [batch_size,1], which is what scatter_nd
-    # needs
-    scatter_indices = tf.where(cond)
-    scope = get_current_scope()
+    # shape of indices from where() is a tuple (indices, )
+    gather_indices = torch.where(cond)[0]
 
     def _update_subset():
-        gather_indices = tf.squeeze(scatter_indices, 1)
         selected_args = _gather_nest(args, gather_indices)
         selected_kwargs = _gather_nest(kwargs, gather_indices)
-        with tf.name_scope(scope):
-            # tf.case loses the original name scope. Need to restore it.
-            updates = func(*selected_args, **selected_kwargs)
-        return tf.nest.map_structure(
-            lambda tgt, updt: tf.tensor_scatter_nd_update(
-                tgt, scatter_indices, updt), target, updates)
-
-    total = tf.shape(cond)[0]
-    n = tf.shape(scatter_indices)[0]
-    return tf.case(((n == 0, lambda: target),
-                    (n == total, lambda: func(*args, **kwargs))),
-                   default=_update_subset)
-
-
-def run_if(cond, func):
-    """Run a function if `cond` Tensor is True.
-
-    This function is useful for conditionally executing a function only when
-    a condition given by a tf Tensor is True. It is equivalent to the following
-    code if `cond` is a python bool value:
-    ```python
-    if cond:
-        func()
-    ```
-    However, when `cond` is tf bool scalar tensor, the above code does not
-    always do what we want because tensorflow does not allow bool scalar tensor
-    to be used in the same way as python bool. So we have to use tf.cond to do
-    the job.
-
-    Args:
-        cond (tf.Tensor): scalar bool Tensor
-        func (Callable): function to be run
-    Returns:
-        None
-    """
-    scope = get_current_scope()
-
-    def _if_true():
-        # The reason of this line is that inside tf.cond, somehow
-        # get_current_scope() is '', which makes operations and summaries inside
-        # func unscoped. We need this line to restore the original name scope.
-        with tf.name_scope(scope):
-            func()
-            return tf.constant(True)
-
-    tf.cond(cond, _if_true, lambda: tf.constant(False))
+        updates = func(*selected_args, **selected_kwargs)
+
+        def _update(tgt, updt):
+            scatter_indices = torch.reshape(
+                gather_indices,
+                list(gather_indices.shape) +
+                [-1] * (updt.ndim - gather_indices.ndim))
+            scatter_indices = scatter_indices.expand_as(updt)
+            return tgt.scatter(0, scatter_indices, updt)
```

Comment:
I wonder whether this more readable way work (whether gradient is correctly handled):
```python
ret = tgt.clone()
ret[gather_indices] = updt
return ret
```

Reply:
For some pytorch operators, in-place operations will throw errors.